tempesta_fw/tls.c

```diff
@@ -379,23 +388,92 @@ tfw_tls_encrypt(struct sock *sk, struct sk_buff *skb, unsigned int limit)
 
 	if (likely(sgt.nents <= AUTO_SEGS_N)) {
 		sgt.sgl = sg;
+		out_sgt.sgl = out_sg;
+		pages = auto_pages;
 	} else {
 		sgt.sgl = kmalloc(sizeof(struct scatterlist) * sgt.nents,
 				  GFP_ATOMIC);
 		if (!sgt.sgl) {
 			T_WARN("cannot alloc TLS encryption scatter list.\n");
-			return -ENOMEM;
+			r = -ENOMEM;
+			goto err_sgt_alloc;
+		}
+		out_sgt.sgl = kmalloc(sizeof(struct scatterlist) * out_sgt.nents,
+				      GFP_ATOMIC);
+		if (!out_sgt.sgl) {
+			T_WARN("cannot alloc TLS encryption scatter list.\n");
+			r = -ENOMEM;
+			goto err_out_sgt_alloc;
+		}
+		pages = kmalloc(sizeof(struct page *) * out_sgt.nents,
+				GFP_ATOMIC);
+		if (!pages) {
+			T_WARN("cannot alloc memory.\n");
+			r = -ENOMEM;
+			goto err_pages_alloc;
 		}
 	}
 	sg_init_table(sgt.sgl, sgt.nents);
+	sg_init_table(out_sgt.sgl, out_sgt.nents);
 
-	for (next = skb, frags = 0; ; ) {
+	for (next = skb, frags = 0, out_frags = 0, elt = 0; ; ) {
+		unsigned int head_data_len = skb_headlen(next);
 		/*
 		 * skb data and tails are already adjusted above,
 		 * so use zero offset and skb->len.
 		 */
 		r = skb_to_sgvec(next, sgt.sgl + frags, 0, next->len);
 
+		/* TODO: process of SKBTX_ZEROCOPY_FRAG for MSG_ZEROCOPY */
+		if (skb_shinfo(next)->tx_flags & SKBTX_SHARED_FRAG) {
+			if (head_data_len) {
+				sg_set_buf(out_sgt.sgl + out_frags, next->data,
+				           head_data_len);
+				out_frags++;
+			}
+
+			remain = offset = 0;
+			for (i = 0; i < skb_shinfo(next)->nr_frags; i++) {
+				skb_frag_t *f = &skb_shinfo(next)->frags[i];
+				unsigned int size;
+				struct page *p;
+
+				size = skb_frag_size(f);
+				if (remain < size) {
+					int order = get_order(size);
+					p = alloc_pages(GFP_ATOMIC, order);
+					if (!p) {
+						r = -ENOMEM;
+						goto out;
+					}
+					remain = (1 << order) * PAGE_SIZE;
+					offset = 0;
+				} else {
+					skb_frag_t *prev_f;
+
+					prev_f = &skb_shinfo(next)->frags[i - 1];
+					p = skb_frag_page(prev_f);
+					get_page(p);
+				}
+				pages[elt++] = skb_frag_page(f);
+				sg_set_page(out_sgt.sgl + out_frags, p, size,
+					    offset);
+				__skb_fill_page_desc(next, i, p, offset, size);
+				remain -= size;
+				offset += size;
+				out_frags++;
+			}
+			if (out_frags > 0)
+				sg_mark_end(&out_sgt.sgl[out_frags - 1]);
+			skb_shinfo(next)->tx_flags &= ~SKBTX_SHARED_FRAG;
+		} else {
+			r = skb_to_sgvec(next, out_sgt.sgl + out_frags, 0,
+				         next->len);
+			if (r <= 0)
+				goto out;
+			out_frags += r;
+		}
```

Comment:
`tfw_tls_encrypt()` is already overloaded function and the code just prepares memory for encrypted data, so it could make sense to move it to separate function in `ss_skb`. However, I'm not sure how many local variables are used - if the function must accept more than standard 6 arguments, them let's leave with the current code.